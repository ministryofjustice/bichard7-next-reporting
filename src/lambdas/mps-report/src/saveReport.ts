import { PostgresGateway } from "@bichard/postgres-gateway"

export default async (gateway: PostgresGateway, areaCode: string, report: string) => {
    const updateQuery = `
DO $$
    BEGIN
        IF EXISTS (SELECT 1 FROM br7own.work_allocation_report WHERE area_code = $1) THEN
            BEGIN
                UPDATE br7own.work_allocation_report
                SET
                    report=cast($2 as bytea),
                    report_timestamp=NOW()
                WHERE
                    area_code = $1;
            END;
        ELSE
            BEGIN
                INSERT INTO br7own.work_allocation_report
                VALUES($1, cast($2 as bytea), NOW());
            END;
        end if;
    END
$$
    `
    const result = await gateway.getResult(updateQuery, [areaCode, report])

    return result
}
