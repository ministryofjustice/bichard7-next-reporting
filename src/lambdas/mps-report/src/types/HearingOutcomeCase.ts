import type OffenceDetails from "./OffenceDetails"
import type XmlStringCell from "./XmlStringCell"

export default interface HearingOutcomeCase {
  PTIURN: XmlStringCell
  HearingDefendant: {
    ArrestSummonsNumber: XmlStringCell
    Offence: OffenceDetails | [OffenceDetails]
    DefendantDetail: {
      BirthDate: XmlStringCell
    }
  }
}
