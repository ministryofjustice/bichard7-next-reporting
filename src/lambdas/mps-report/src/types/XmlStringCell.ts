export function getText(object: {_text:string | undefined }){
    if(!object || !object._text) {
        return ""
    }
    return object._text
}

export default interface XmlStringCell{
    _text: string
}