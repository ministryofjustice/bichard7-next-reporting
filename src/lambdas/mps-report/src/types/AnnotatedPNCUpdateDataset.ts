import HearingOutcomeCase from "./HearingOutcomeCase"

export default interface AnnotatedPNCUpdateDataset {
  AnnotatedPNCUpdateDataset: {
    PNCUpdateDataset: {
      AnnotatedHearingOutcome: {
        HearingOutcome: {
          Case: HearingOutcomeCase
        }
      }
    }
  }
}
