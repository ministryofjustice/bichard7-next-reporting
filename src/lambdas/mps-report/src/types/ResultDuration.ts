import XmlStringCell from "./XmlStringCell";

export default interface ResultDuration {
    DurationUnit: XmlStringCell
    DurationLength: XmlStringCell
}