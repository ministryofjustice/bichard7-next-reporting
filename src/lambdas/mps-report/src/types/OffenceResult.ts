import ResultDuration from "./ResultDuration"
import ResultQualifier from "./ResultQualifier"
import XmlStringCell from "./XmlStringCell"

export default interface OffenceResult {
  CJSresultCode: XmlStringCell
  Duration: [ResultDuration] | ResultDuration
  AmountSpecifiedInResult: [XmlStringCell] | XmlStringCell
  ResultVariableText: XmlStringCell
  ResultQualifierVariable: [ResultQualifier] | ResultQualifier
}
