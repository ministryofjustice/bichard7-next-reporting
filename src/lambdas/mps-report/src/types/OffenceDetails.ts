import OffenceReason from "./OffenceReason"
import OffenceResult from "./OffenceResult"
import XmlStringCell from "./XmlStringCell"

export function isMultiple<T>(obj: T | [T]): obj is [T]{
    return (obj as [T]).length > 0
}

export default interface OffenceDetails {
  ActualOffenceStartDate: {
    StartDate: XmlStringCell
  }
  CriminalProsecutionReference: OffenceReason
  LocationOfOffence: XmlStringCell
  Result: [OffenceResult] | OffenceResult
}
