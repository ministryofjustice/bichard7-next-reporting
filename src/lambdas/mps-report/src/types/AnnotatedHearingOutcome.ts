import OffenceDetails from "./OffenceDetails"
import XmlStringCell from "./XmlStringCell"

export default interface AnnotatedHearingOutcome {
  AnnotatedHearingOutcome: {
    HearingOutcome: {
      Case: {
        PTIURN: XmlStringCell
        HearingDefendant: {
          ArrestSummonsNumber: XmlStringCell
          Offence: OffenceDetails | [OffenceDetails]
          DefendantDetail: {
            BirthDate: XmlStringCell
          }
        }
      }
    }
  }
}
