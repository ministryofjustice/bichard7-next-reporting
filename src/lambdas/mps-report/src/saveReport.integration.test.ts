import { PostgresGateway } from "@bichard/postgres-gateway"
import config from "./lib/config"
import saveReport from "./saveReport"
import { isError } from "@bichard/types"
var JSZip = require("jszip")

describe("GenerateReport", () => {
  let gateway: PostgresGateway

  beforeAll(async () => {
    gateway = new PostgresGateway(config.database)
  })

  beforeEach(async () => {
    await gateway.getResult(`DELETE FROM br7own.work_allocation_report`)
  })

  afterAll(async () => {
    await gateway.close()
  })

  it("should save a zipped report to the DB", async () => {
    const reportToSave = "Best,Report,In,The,World"
    const saveResult = await saveReport(gateway, "01", reportToSave)
    if (isError(saveResult)) {
      expect(isError(saveResult)).toBeFalsy
      return
    }

    const query = "SELECT * FROM br7own.work_allocation_report"
    const result = await gateway.getResult(query)
    if (isError(result)) {
      expect(isError(result)).toBeFalsy
      return
    }
    expect(result).toHaveLength(1)
    const zipData = await JSZip.loadAsync(result[0].report)
    const zipFiles = zipData.file(/Area01\.\d{10}\.csv/)
    expect(zipFiles).toHaveLength(1)
    const content = await zipFiles[0].async("string")
    expect(content).toEqual(reportToSave)
  })
})
