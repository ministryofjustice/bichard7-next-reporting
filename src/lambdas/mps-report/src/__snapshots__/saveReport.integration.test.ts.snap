// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerateReport should save a report to the DB 1`] = `"Best,Report,In,The,World"`;
